sklearn/cluster/_hierarchical_fast.pyx

```diff
@@ -29,15 +25,17 @@ from numpy.math cimport INFINITY
 ###############################################################################
 # Utilities for computing the ward momentum
 
-def compute_ward_dist(cnp.ndarray[DOUBLE, ndim=1, mode='c'] m_1,
-                      cnp.ndarray[DOUBLE, ndim=2, mode='c'] m_2,
-                      cnp.ndarray[INTP, ndim=1, mode='c'] coord_row,
-                      cnp.ndarray[INTP, ndim=1, mode='c'] coord_col,
-                      cnp.ndarray[DOUBLE, ndim=1, mode='c'] res):
-    cdef INTP size_max = coord_row.shape[0]
-    cdef INTP n_features = m_2.shape[1]
-    cdef INTP i, j, row, col
-    cdef DOUBLE pa, n
+cpdef void compute_ward_dist(
+    const cnp.float64_t[::1] m_1,
+    const cnp.float64_t[:, ::1] m_2,
+    const cnp.intp_t[::1] coord_row,
+    const cnp.intp_t[::1] coord_col,
+    cnp.float64_t[::1] res
+) nogil:
```

Comment:
@jjerphan I don't get the need to make it a cpdef nogil. This function is only used on the python side. A cpdef creates 2 versions of the function in the binary, one of them being never used.